libevmasm/PeepholeOptimiser.cpp

```diff
@@ -428,6 +428,47 @@ struct TruthyAnd: SimplePeepholeOptimizerMethod<TruthyAnd>
 	}
 };
 
+// (a == true) -> a and (true == a) -> a
+struct TruthyBooleanComparison: SimplePeepholeOptimizerMethod<TruthyBooleanComparison>{
+	static bool applySimple(
+		AssemblyItem const& _pushOrDupN,
+		AssemblyItem const& _iszero,
+		AssemblyItem const& _iszero1,
+		AssemblyItem const& _pushOrDupN1,
+		AssemblyItem const& _iszero2,
+		AssemblyItem const& _iszero3,
+		AssemblyItem const& _sub,
```

Comment:
> One thing I noticed is that the equality is checked within an if condition is with `SUB` (since it expects 0 for `JUMPI`?), but if we simply do ` return a == true;` the equality here is checked with `EQ`. 
>
> I guess that means we have to optimize for both these cases, or am I misunderstanding something here?

The code generated for equality is `EQ` and for inequality `EQ ISZERO`: https://github.com/ethereum/solidity/blob/b4b5f7b355d46e8f286926d9b0da613ecc7a09dc/libsolidity/codegen/ExpressionCompiler.cpp#L2336-L2339

The thing is that the condition `JUMPI` checks is actually the `else` condition. I.e. we just continue past `JUMPI` when the condition is satisfied and we jump when it's not. So `EQ` for equality gets negated into `EQ ISZERO` and that gets turned into `SUB` by the `EqIsZeroJumpI` rule: https://github.com/ethereum/solidity/blob/b4b5f7b355d46e8f286926d9b0da613ecc7a09dc/libevmasm/PeepholeOptimiser.cpp#L304-L315

The rule is part of the peephole optimizer so I think you can just assume it will always get applied and only handle the code after its application. So not need to handle `EQ` separately if you only want to handle `if`.

On the other hand, it may still appear in other constructs. For example:
```solidity
    function id(bool a) external returns (bool) {
        return a == true;
    }
```
looks like this unoptimized:
```
      0x01
      iszero
      iszero
      dup3
      iszero
      iszero
      eq
      swap1
      pop
      swap2
      swap1
      pop
      jump
```

though currently it gets optimized into this eventually:
```
      iszero
      iszero
      0x01
      eq
      swap1
      jump
```